Bug: deleting a class which has a running classification leads to a panic

Using the classification setup from the weaviate-javascript-client e2e test, the following panic occurs:

```
panic: trying to delete nil node 0 of parent 0 (0.094482)


goroutine 327 [running]:
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*binarySearchNodeGeneric).delete(0x0, 0x0, 0x0, 0xc0009d5a40)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/binary_search_tree.go:232 +0x3ee
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*binarySearchNodeGeneric).delete(0xc0009d5a40, 0x0, 0x0, 0xc0009d59e0)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/binary_search_tree.go:279 +0x295
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*binarySearchNodeGeneric).delete(0xc0009d59e0, 0x0, 0x0, 0xc0009d5a80)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/binary_search_tree.go:282 +0x2bc
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*binarySearchTreeGeneric).delete(0xc0004e67a0, 0x0, 0x0)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/binary_search_tree.go:78 +0x91
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).findAndConnectNeighbors(0xc000a32000, 0xc000be9aa0, 0x3, 0xc000c0af00, 0x12c, 0x12c, 0x1, 0x1, 0x0, 0x0, ...)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/index.go:464 +0x701
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).insert(0xc000a32000, 0xc000be9aa0, 0xc000c0af00, 0x12c, 0x12c, 0xc000b56000, 0xc000b2f501)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/index.go:335 +0x2ea
github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).Add(0xc000a32000, 0x0, 0xc000c0af00, 0x12c, 0x12c, 0xc000ae1e88, 0x10075ab)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/vector/hnsw/index.go:267 +0xc1
github.com/semi-technologies/weaviate/adapters/repos/db.(*Shard).updateVectorIndex(0xc000a83ec0, 0xc000c0af00, 0x12c, 0x12c, 0x0, 0xc000b56100, 0x0, 0xc000413620, 0xc000c50580)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/shard_write_put.go:65 +0x7f
github.com/semi-technologies/weaviate/adapters/repos/db.(*objectsBatcher).storeSingleObjectInAdditionalStorage(0xc000c50540, 0x1edb0c0, 0xc000b52720, 0xc000b3cd10, 0x0, 0xc00062f100, 0x0, 0x0)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/shard_write_batch_objects.go:209 +0xa6
github.com/semi-technologies/weaviate/adapters/repos/db.(*objectsBatcher).storeAdditionalStorage.func1(0xc0004515f0, 0xc000c50540, 0x1edb0c0, 0xc000b52720, 0xc000b3cd10, 0x0, 0x0, 0x0, 0x0)
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/shard_write_batch_objects.go:179 +0xbf
created by github.com/semi-technologies/weaviate/adapters/repos/db.(*objectsBatcher).storeAdditionalStorage
        /Users/etiennedilocker/go/src/github.com/semi-technologies/weaviate/adapters/repos/db/shard_write_batch_objects.go:177 +0x211
```

--- a/usecases/classification/classifier_run.go
+++ b/usecases/classification/classifier_run.go
@@ -20,6 +20,7 @@ import (
 	"github.com/go-openapi/strfmt"
 	"github.com/pkg/errors"
 	"github.com/semi-technologies/weaviate/entities/models"
+	"github.com/semi-technologies/weaviate/entities/schema"
 	"github.com/semi-technologies/weaviate/entities/search"
 	"github.com/sirupsen/logrus"
 )
@@ -33,8 +34,10 @@ type classifyItemFn func(item search.Result, itemIndex int,
 
 func (c *Classifier) run(params models.Classification,
 	filters filters) {
-	ctx, cancel := contextWithTimeout(30 * time.Second)
+	ctx, cancel := contextWithTimeout(30 * time.Minute)
 	defer cancel()
+
+	go c.monitorClassification(ctx, cancel, schema.ClassName(params.Class))
 
 	c.logBegin(params, filters)
 	unclassifiedItems, err := c.vectorRepo.GetUnclassified(ctx,
@@ -57,13 +60,31 @@ func (c *Classifier) run(params models.Classification,
 		return
 	}
 
-	params, err = c.runItems(classifyItem, params, filters, unclassifiedItems)
+	params, err = c.runItems(ctx, classifyItem, params, filters, unclassifiedItems)
 	if err != nil {
 		c.failRunWithError(params, err)
 		return
 	}
 
 	c.succeedRun(params)
+}
+
+func (c *Classifier) monitorClassification(ctx context.Context, cancelFn context.CancelFunc,
+	className schema.ClassName) {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			schema := c.schemaGetter.GetSchemaSkipAuth()
+			class := schema.FindClassByName(className)
+			if class == nil {
+				cancelFn()
+				return
+			}
+		}
+	}
 }
 
 func (c *Classifier) prepareRun(params models.Classification, filters filters,
@@ -94,7 +115,7 @@ func (c *Classifier) prepareRun(params models.Classification, filters filters,
 
 // runItems splits the job list into batches that can be worked on parallelly
 // depending on the available CPUs
-func (c *Classifier) runItems(classifyItem classifyItemFn, params models.Classification, filters filters,
+func (c *Classifier) runItems(ctx context.Context, classifyItem classifyItemFn, params models.Classification, filters filters,
 	items []search.Result) (models.Classification, error) {
 	workerCount := runtime.GOMAXPROCS(0)
 	if len(items) < workerCount {
@@ -103,7 +124,7 @@ func (c *Classifier) runItems(classifyItem classifyItemFn, params models.Classif
 
 	workers := newRunWorkers(workerCount, classifyItem, params, filters, c.vectorRepo)
 	workers.addJobs(items)
-	res := workers.work()
+	res := workers.work(ctx)
 
 	params.Meta.Completed = strfmt.DateTime(time.Now())
 	params.Meta.CountSucceeded = res.successCount

--- a/usecases/classification/classifier_run_worker.go
+++ b/usecases/classification/classifier_run_worker.go
@@ -12,6 +12,7 @@
 package classification
 
 import (
+	"context"
 	"sync"
 	"sync/atomic"
 
@@ -37,10 +38,17 @@ func (w *runWorker) addJob(job search.Result) {
 	w.jobs = append(w.jobs, job)
 }
 
-func (w *runWorker) work(wg *sync.WaitGroup) {
+func (w *runWorker) work(ctx context.Context, wg *sync.WaitGroup) {
 	defer wg.Done()
 
 	for i, item := range w.jobs {
+		// check if the whole classification operation has been cancelled
+		// if yes, then abort the classifier worker
+		if err := ctx.Err(); err != nil {
+			w.ec.add(err)
+			atomic.AddInt64(w.errorCount, 1)
+			break
+		}
 		originalIndex := (i * w.workerCount) + w.id
 		err := w.classify(item, originalIndex, w.params, w.filters, w.batchWriter)
 		if err != nil {
@@ -106,13 +114,13 @@ func (ws *runWorkers) addJobs(jobs []search.Result) {
 	}
 }
 
-func (ws *runWorkers) work() runWorkerResults {
+func (ws *runWorkers) work(ctx context.Context) runWorkerResults {
 	ws.batchWriter.Start()
 
 	wg := &sync.WaitGroup{}
 	for _, worker := range ws.workers {
 		wg.Add(1)
-		go worker.work(wg)
+		go worker.work(ctx, wg)
 	}
 
 	wg.Wait()

